data_processors/pipeline/lambdas/orchestrator.py

```diff
@@ -71,6 +73,153 @@ def update_step(wfr_id, wfv_id, wfr_event, context):
     return None
 
 
+def get_germline_runs(sequence_run: SequenceRun):
+    # query for Workflows associated with this SequenceRun
+    # that are GERMLINE workflows
+    # and that don't have an end status (yet)
+    workflows = Workflow.objects.filter(
+        sequence_run=sequence_run,
+        type_name__iexact=WorkflowType.GERMLINE.value.lower())
+    running = list()
+    ended = list()
+
+    for wf in workflows:
+        if wf.end_status:  # TODO: check if that works
+            ended.append(wf)
+        else:
+            running.append(wf)
+
+    return running, ended
+
+
+def get_library_id_from_workflow(workflow: Workflow):
+    # extract library ID from Germline WF sample_name
+    # TODO: is there a better way? Could use the fastq_list_row entries of the workflow 'input'
+
+    # remove the first part (sample ID) from the sample_name to get the library ID
+    # NOTE: this may not be exactly the library ID used in the Germline workflow (stripped off _topup/_rerun), but
+    #       for our use case that does not matter, as we are merging all libs from the same subject anyway
+    return '_'.join(workflow.sample_name.split('_')[1:])
+
+
+def get_subjects_from_runs(workflows: list) -> list:
+    subjects = set()
+    for workflow in workflows:
+        # use workflow helper class to extract sample/library ID from workflow
+        library_id = get_library_id_from_workflow(workflow)
+        if library_id:
+            # use metadata helper class to get corresponding subject ID
+            subject_id = LabMetadata.objects.get(library_id=library_id).subject_id
+            if subject_id:
+                subjects.add(subject_id)
+            else:
+                raise ValueError(f"No subject for library {library_id}")
+        else:
+            raise ValueError(f"Could not extract LibraryID from workflow {workflow.wfr_id}")
+
+    return list(subjects)
+
+
+def extract_sample_library_ids(fastq_list_rows: list[FastqListRow]):
+    samples = set()
+    libraries = set()
+
+    for row in fastq_list_rows:
+        libraries.add(row.rglb)
+        samples.add(row.rgsm)
+
+    return list(samples), list(libraries)
+
+
+def prepare_tumor_normal_jobs(subjects: list[str]) -> list:
+    jobs = list()
+    for subject in subjects:
+        job = create_tn_job(subject)
+        if job:
+            jobs.append(job)
+        else:
+            logger.debug(f"Not running T/N workflow for {subject}. Conditions not met.")
+
+    return jobs
+
+
+def create_tn_job(subject_id: str) -> dict:
+    # TODO: could query for all records in one go and then filter locally
+    # records = LabMetadata.objects.filter(subject_id=subject_id)
+
+    # extract tumor/normal pairs for a subject (taking into account Type & Phenotype)
+    tumor_records = LabMetadata.objects.filter(
```

Comment:
I'm not familiar / across the LabMetadata models but I assume this is where the magic is in terms of collecting a list of 'records' containing WGS records that can be mapped to the fastq list rows objects in the lines below?

Reply:
Yup. Here we first query for metadata records (WGS, tumor samples) of a single subject. (same for normals)
Then further down we try to find FastqListRow entries for those metadata records (based on the library ID).

Still needs to deal with topup/rerun cases, but if we have at least one tumor and one normal record/FastqListRow, we proceed to preparing the T/N job.